invenio_app_ils/records/api.py

```diff
@@ -525,3 +550,80 @@ def create(cls, data, id_=None, **kwargs):
             )
         }
         return super(Series, cls).create(data, id_=id_, **kwargs)
+
+
+class DocumentRequest(IlsRecord):
+    """DocumentRequest record class."""
+
+    _pid_type = DOCUMENT_REQUEST_PID_TYPE
+    _schema = "document_requests/document_request-v1.0.0.json"
+
+    _document_resolver_path = (
+        "{scheme}://{host}/api/resolver/document-requests/"
+        "{document_request_pid}/document"
+    )
+
+    @classmethod
+    def _validate_fulfilled(cls, document_pid):
+        """Validate data for fulfilled state."""
+        # Fulfilled requests must have a document
+        if not document_pid:
+            raise DocumentRequestError(
+                "State cannot be 'FULFILLED' without a document"
+            )
+        try:
+            document = Document.get_record_by_pid(document_pid)
+            document = document.replace_refs()
+            request = document["request"]
+            if request:
+                # The document cannot already have a request
+                raise DocumentRequestError(
+                    "Document with PID {} already has a request".format(
+                        document_pid
+                    )
+                )
+        except PIDDoesNotExistError:
+            # Invalid document_pid
+            raise DocumentRequestError(
+                "State cannot be 'FULFILLED' because a document with "
+                "PID {} doesn't exist".format(document_pid)
+            )
+
+    @classmethod
+    def validate_create(cls, data):
+        """Validate document request data."""
+        if data.get("state", "") == "FULFILLED":
+            cls._validate_fulfilled(data.get("document_pid", None))
+
+    def validate_patch(self, ops):
+        """Validate document request data."""
+        for op in ops:
+            if "state" in op["path"] and op["value"] == "FULFILLED":
+                # Ensure there is also a document_pid in the ops
+                found_document_pid = False
+                for op in ops:
+                    if "document_pid" in op["path"]:
+                        self._validate_fulfilled(document_pid=op["value"])
+                        found_document_pid = True
+                if not found_document_pid:
```

Comment:
I am not sure I understand fully this case. Is it happening when it is already in FULFILLED state?

Reply:
This case would happen if you were to patch the `state` and change it to `FULFILLED` but not provide a `document_pid`. In that case, we validate it it with `document_pid=None` which will raise an exception.

Reply:
Can I suggest something, again for next PR, just to improve readibility? (even if I am not sure it does improve :))

```
fulfilledStateOp = next(iter(filter(ops, lambda op: op.path == 'state' and op.value == 'FULFILLED'), None))
documentPidOp = next(iter(filter(ops, lambda op: path == 'document_pid'), None))
if fulfilledStateOp:
    self._validate_fulfilled(document_pid=documentPidOp["value"])
```

I am not a big fan of `next(iter))` but it is one line instruction to get the element (not a list) or `None`.
WDYT?

Maybe you can find a better solution too...

Reply:
For me that looks good!

Another suggestion:
```python
fulfilled_state = any(op['path'] == '/state' and op['value'] == 'FULFILLED' for op in ops)
document_pid = next((op['value'] for op in ops if op['path'] == '/document_pid'), None)
if fulfilled_state:
    self._validate_fulfilled(document_pid=document_pid)
```